{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of [Hui-Ling Yen]'s *A/California/04/2009* sequences and design of synonymously \"barcoded\" variants\n",
    "This [iPython notebook](http://ipython.org/notebook.html) describes the analysis of pdmH1N1 influenza gene sequences and the design of synonymous \"barcodes\" which are intended to be selectively neutral but allow for distinguishing between variants by sequencing.\n",
    "\n",
    "This analysis is performed by [Jesse Bloom](http://research.fhcrc.org/bloom/en.html) to contribute to a larger project being led by [Hui-Ling Yen].\n",
    "\n",
    "## Compilation of a set of naturally occuring sequences\n",
    "As a point of comparison, we assemble a set of naturally occurring sequences. To do this, on September-5-2015, I downloaded from the [Influenza Virus Resource](http://www.ncbi.nlm.nih.gov/genomes/FLU/FLU.html) all full-length human influenza coding DNA sequences for human H1N1 annotated as *pandemic H1N1*. Here are some details on that sequence set:\n",
    "\n",
    "[Hui-Ling Yen]: http://sph.hku.hk/en/about-us/faculty-and-staff/academic-staff/yen,-hui-ling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read a total of 68816 sequences from ./sequences/all_pdmH1N1_seqs.fasta\n",
      "For PB2, found 5114 sequences\n",
      "For PB1, found 5125 sequences\n",
      "For PA, found 5358 sequences\n",
      "For HA, found 11321 sequences\n",
      "For NP, found 5457 sequences\n",
      "For NA, found 8990 sequences\n",
      "For M1, found 7608 sequences\n",
      "For M2, found 7051 sequences\n",
      "For NS1, found 5907 sequences\n",
      "For NS2, found 5537 sequences\n"
     ]
    }
   ],
   "source": [
    "seqdir = './sequences' \n",
    "allseqs = '%s/all_pdmH1N1_seqs.fasta' % seqdir\n",
    "cdnas = ['PB2', 'PB1', 'PA', 'HA', 'NP', 'NA', 'M1', 'M2', 'NS1', 'NS2'] # we analyze these cDNAs\n",
    "ignored_cdnas = ['PA-X', 'PB1-F2'] # we ignore these cDNAs\n",
    "\n",
    "import Bio.SeqIO\n",
    "seqs = [seq for seq in Bio.SeqIO.parse(allseqs, 'fasta')]\n",
    "print('Read a total of %d sequences from %s' % (len(seqs), allseqs))\n",
    "seqs_by_cdna = dict([(cdna, []) for cdna in cdnas])\n",
    "for seq in seqs:\n",
    "    cdna = seq.description.strip().split()[-1]\n",
    "    if cdna in seqs_by_cdna:\n",
    "        seqs_by_cdna[cdna].append(seq)\n",
    "    elif cdna not in ignored_cdnas:\n",
    "        print(\"Unrecognized cDNA of %s in sequence header %s\" % (cdna, seq.description))\n",
    "for cdna in cdnas:\n",
    "    print(\"For %s, found %d sequences\" % (cdna, len(seqs_by_cdna[cdna])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reference sequences for experimental work from [Hui-Ling Yen]'s plasmids\n",
    "Our experiments will utilize pdmH1N1 sequences cloned into reverse genetics plasmids by [Hui-Ling Yen].\n",
    "\n",
    "She has provided sequences for reverse-genetics plasmids encoding the genes for the strains *A/California/04/2009* and *A/Hong Kong/415742/2009*, which are both pdmH1N1. It appears that the references for these plasmids are [Yen, PNAS, 108:14264](http://www.pnas.org/content/108/34/14264.full) and [Wong, J Virology, 86:10558](http://jvi.asm.org/content/86/19/10558.abstract), respectively. From those references, it is clear that both viruses can be rescued from the reverse-genetics plasmids. According to [Hui-Ling Yen]'s e-mail, the coding sequences for the two viruses are identical for *PB1*, *NA*, *M*, and *NS* -- but differ for *PB2*, *PA*, *HA*, and *NP*. She also noted that there are four non-synonymous mutations in *PB1* (perhaps relative to the Genbank reference, I wasn't clear on this point?) that should be examined. Her rationale for preferring to use the *A/California/04/2009* virus over the *A/Hong Kong/415742/2009* is that the former is a more commonly used reference strain for studies, whereas the latter is less used and has some differences relative to many other pdmH1N1 strains. I had heard anecdotally that some *A/California/04/2009* strains are difficult to rescue and/or propagate in tissue-culture, but based on her prior work with this virus it doesn't seem to be a big problem for her particular variant.\n",
    "\n",
    "In the analysis that follows, I will **only be examining the coding sequences, not the noncoding regions.** The reason is that the noncoding regions in Genbank sequences are often unreliable, and I'm going to assume that the noncoding regions in [Hui-Ling Yen]'s plasmids are fine since she has successfully used them in the past.\n",
    "\n",
    "First, we read in the sequences that [Hui-Ling Yen] provided for the inserts in her reverse-genetics plasmids:\n",
    "\n",
    "[Hui-Ling Yen]: http://sph.hku.hk/en/about-us/faculty-and-staff/academic-staff/yen,-hui-ling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 8 sequences from ./sequences/plasmid_sequences_20150904.txt. These are:\n",
      "\tCA04-PB2 plasmid\n",
      "\tpYW472-PB1 plasmid\n",
      "\tCA04-PA plasmid\n",
      "\tCA04-HA plasmid\n",
      "\tCA04-NP plasmid\n",
      "\tpYW476-NA plasmid\n",
      "\tpYW477-M plasmid\n",
      "\tpYW478-NS plasmid\n",
      "Found a reading frame of 2277 nucleotides (759 codons) for PB2\n",
      "Found a reading frame of 2271 nucleotides (757 codons) for PB1\n",
      "Found a reading frame of 2148 nucleotides (716 codons) for PA\n",
      "Found a reading frame of 1698 nucleotides (566 codons) for HA\n",
      "Found a reading frame of 1494 nucleotides (498 codons) for NP\n",
      "Found a reading frame of 1407 nucleotides (469 codons) for NA\n",
      "Found a reading frame of 756 nucleotides (252 codons) for M1\n",
      "Found a reading frame of 291 nucleotides (97 codons) for M2\n",
      "Found a reading frame of 657 nucleotides (219 codons) for NS1\n",
      "Found a reading frame of 363 nucleotides (121 codons) for NS2\n"
     ]
    }
   ],
   "source": [
    "plasmidseqs = '%s/plasmid_sequences_20150904.txt' % seqdir\n",
    "plasmids = [seq for seq in Bio.SeqIO.parse(plasmidseqs, 'fasta')]\n",
    "print(\"Read %d sequences from %s. These are:\\n\\t%s\" % (len(plasmids), plasmidseqs, '\\n\\t'.join([seq.description for seq in plasmids])))\n",
    "\n",
    "# Now get the plasmid coding sequences into plasmid_cdnas; this requires trimming the noncoding regions\n",
    "plasmid_cdnas = {}\n",
    "splicesites = {'M2':(26, 715), 'NS2':(30, 503)} # tuples give last nt of exon 1 & first of exon 2 in numbering starting with 1 at ATG\n",
    "for cdna in cdnas:\n",
    "    plasmidseq = [seq.seq for seq in plasmids if (cdna in seq.description) or (cdna in ['M1', 'NS1', 'M2', 'NS2'] and cdna[ : -1] in seq.description)]\n",
    "    assert len(plasmidseq) == 1, \"Failed to find a single plasmid sequence for %s\" % cdna\n",
    "    plasmidseq = plasmidseq[0]\n",
    "    orfstart = str(plasmidseq).index('ATG')\n",
    "    if cdna == 'NP': # starts with second ATG in RNA\n",
    "        orf = plasmidseq[str(plasmidseq).index('ATG', orfstart + 1) : ]\n",
    "    elif cdna not in ['M2', 'NS2']: # starts with first ATG in RNA\n",
    "        orf = plasmidseq[orfstart : ]\n",
    "    else: # things are bit more complicated for the spliced genes M2 and NS2\n",
    "        (splice5, splice3) = splicesites[cdna]\n",
    "        exon1 = plasmidseq[orfstart : orfstart + splice5]\n",
    "        exon2 = plasmidseq[orfstart + splice3 - 1 : ]\n",
    "        orf = exon1 + exon2\n",
    "    prot = orf.translate(to_stop=True)\n",
    "    orf = orf[ : 3 * len(prot)]\n",
    "    plasmid_cdnas[cdna] = orf\n",
    "    print(\"Found a reading frame of %s nucleotides (%d codons) for %s\" % (len(orf), len(orf) // 3, cdna))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison of reference sequences to those in Genbank\n",
    "We now compare the reference sequences from the plasmids to all sequences present for pandemic H1N1 that we have downloaded from the [Influenza Virus Resource](http://www.ncbi.nlm.nih.gov/genomes/FLU/FLU.html). Since the sequences should all be nearly identical, we don't perform true alignments -- we just make sure they are length-matched as there are no indels during the evolution of pdmH1N1. Below are the results of this analysis, which show that while not all of the **nucleotide** sequences are exact matches to *A/California/04/2009* database sequences, all of the **protein** sequences are matches to at least one *A/California/04/2009* sequence. Overall, this seems to indicate that the plasmid sequences are probably okay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "For PB2, 5111 of 5114 sequences aligned gaplessly. The median and minimum nucleotide identities are 0.997 and 1.000\n",
      "The PB2 nucleotide sequence in the plasmid exactly matches 16 known sequences.\n",
      "The PB2 protein sequence in the plasmid exactly matches 1544 known sequences.\n",
      "The exact nucleotide matches for PB2 include A/California/04/2009 sequences:\n",
      "\tcds:ACP41102 A/California/04/2009 2009/04/01 H1N1 PB2\n",
      "\tcds:ACP44157 A/California/04/2009 2009/04/01 H1N1 PB2\n",
      "The exact protein matches for PB2 include A/California/04/2009 sequences:\n",
      "\tcds:ACP41102 A/California/04/2009 2009/04/01 H1N1 PB2\n",
      "\tcds:ACP44157 A/California/04/2009 2009/04/01 H1N1 PB2\n",
      "\n",
      "For PB1, 5101 of 5125 sequences aligned gaplessly. The median and minimum nucleotide identities are 0.997 and 1.000\n",
      "The PB1 nucleotide sequence in the plasmid exactly matches 0 known sequences.\n",
      "The PB1 protein sequence in the plasmid exactly matches 1944 known sequences.\n",
      "There are no exact nucleotide matches for PB1. The closest sequence is:\n",
      "\tcds:ADF27354 A/Hong Kong/415742Md/2009 2009/11/03 H1N1 PB1\n",
      "\tThis sequence has the following nucleotide mutations: C945Y\n",
      "The exact protein matches for PB1 include A/California/04/2009 sequences:\n",
      "\tcds:ACP41103 A/California/04/2009 2009/04/01 H1N1 PB1\n",
      "\tcds:ACT36632 A/California/04/2009 2009/04/01 H1N1 PB1\n",
      "\n",
      "For PA, 5354 of 5358 sequences aligned gaplessly. The median and minimum nucleotide identities are 0.997 and 1.000\n",
      "The PA nucleotide sequence in the plasmid exactly matches 0 known sequences.\n",
      "The PA protein sequence in the plasmid exactly matches 22 known sequences.\n",
      "There are no exact nucleotide matches for PA. The closest sequence is:\n",
      "\tcds:AHZ40833 A/California/07-JRO10/2009 2009/04/09 H1N1 PA\n",
      "\tThis sequence has the following nucleotide mutations: T996C\n",
      "The exact protein matches for PA include A/California/04/2009 sequences:\n",
      "\tcds:ACP41104 A/California/04/2009 2009/04/01 H1N1 PA\n",
      "\tcds:ACP44156 A/California/04/2009 2009/04/01 H1N1 PA\n",
      "\n",
      "For HA, 11272 of 11321 sequences aligned gaplessly. The median and minimum nucleotide identities are 0.994 and 1.000\n",
      "The HA nucleotide sequence in the plasmid exactly matches 4 known sequences.\n",
      "The HA protein sequence in the plasmid exactly matches 8 known sequences.\n",
      "The exact nucleotide matches for HA include A/California/04/2009 sequences:\n",
      "\tcds:ACP41105 A/California/04/2009 2009/04/01 H1N1 HA\n",
      "\tcds:ACQ76318 A/California/04/2009 2009/04/01 H1N1 HA\n",
      "The exact protein matches for HA include A/California/04/2009 sequences:\n",
      "\tcds:ACP41105 A/California/04/2009 2009/04/01 H1N1 HA\n",
      "\tcds:ACQ76318 A/California/04/2009 2009/04/01 H1N1 HA\n",
      "\n",
      "For NP, 5455 of 5457 sequences aligned gaplessly. The median and minimum nucleotide identities are 0.997 and 1.000\n",
      "The NP nucleotide sequence in the plasmid exactly matches 141 known sequences.\n",
      "The NP protein sequence in the plasmid exactly matches 298 known sequences.\n",
      "The exact nucleotide matches for NP include A/California/04/2009 sequences:\n",
      "\tcds:ACP41106 A/California/04/2009 2009/04/01 H1N1 NP\n",
      "\tcds:ACP44151 A/California/04/2009 2009/04/01 H1N1 NP\n",
      "The exact protein matches for NP include A/California/04/2009 sequences:\n",
      "\tcds:ACP41106 A/California/04/2009 2009/04/01 H1N1 NP\n",
      "\tcds:ACP44151 A/California/04/2009 2009/04/01 H1N1 NP\n",
      "\n",
      "For NA, 8989 of 8990 sequences aligned gaplessly. The median and minimum nucleotide identities are 0.997 and 1.000\n",
      "The NA nucleotide sequence in the plasmid exactly matches 164 known sequences.\n",
      "The NA protein sequence in the plasmid exactly matches 341 known sequences.\n",
      "The exact nucleotide matches for NA do not include A/California/04/2009 but do include 164 other sequences including cds:ACP41962 A/Texas/04/2009 2009/04/14 H1N1 NA\n",
      "The exact protein matches for NA include A/California/04/2009 sequences:\n",
      "\tcds:ACP41107 A/California/04/2009 2009/04/01 H1N1 NA\n",
      "\tcds:ACP44158 A/California/04/2009 2009/04/01 H1N1 NA\n",
      "\n",
      "For M1, 7608 of 7608 sequences aligned gaplessly. The median and minimum nucleotide identities are 0.996 and 1.000\n",
      "The M1 nucleotide sequence in the plasmid exactly matches 914 known sequences.\n",
      "The M1 protein sequence in the plasmid exactly matches 5050 known sequences.\n",
      "The exact nucleotide matches for M1 include A/California/04/2009 sequences:\n",
      "\tcds:ACP44152 A/California/04/2009 2009/04/01 H1N1 M1\n",
      "The exact protein matches for M1 include A/California/04/2009 sequences:\n",
      "\tcds:ACP44152 A/California/04/2009 2009/04/01 H1N1 M1\n",
      "\n",
      "For M2, 7032 of 7051 sequences aligned gaplessly. The median and minimum nucleotide identities are 1.000 and 1.000\n",
      "The M2 nucleotide sequence in the plasmid exactly matches 4402 known sequences.\n",
      "The M2 protein sequence in the plasmid exactly matches 5079 known sequences.\n",
      "The exact nucleotide matches for M2 include A/California/04/2009 sequences:\n",
      "\tcds:ACP44153 A/California/04/2009 2009/04/01 H1N1 M2\n",
      "The exact protein matches for M2 include A/California/04/2009 sequences:\n",
      "\tcds:ACP44153 A/California/04/2009 2009/04/01 H1N1 M2\n",
      "\n",
      "For NS1, 5895 of 5907 sequences aligned gaplessly. The median and minimum nucleotide identities are 0.995 and 0.998\n",
      "The NS1 nucleotide sequence in the plasmid exactly matches 0 known sequences.\n",
      "The NS1 protein sequence in the plasmid exactly matches 1058 known sequences.\n",
      "There are no exact nucleotide matches for NS1. The closest sequence is:\n",
      "\tcds:AKQ10385 A/Singapore/DMS962/2009 2009/09/16 H1N1 NS1\n",
      "\tThis sequence has the following nucleotide mutations: T264C\n",
      "The exact protein matches for NS1 include A/California/04/2009 sequences:\n",
      "\tcds:ACP41110 A/California/04/2009 2009/04/01 H1N1 NS1\n",
      "\tcds:ACP44154 A/California/04/2009 2009/04/01 H1N1 NS1\n",
      "\n",
      "For NS2, 5537 of 5537 sequences aligned gaplessly. The median and minimum nucleotide identities are 1.000 and 1.000\n",
      "The NS2 nucleotide sequence in the plasmid exactly matches 3629 known sequences.\n",
      "The NS2 protein sequence in the plasmid exactly matches 4353 known sequences.\n",
      "The exact nucleotide matches for NS2 include A/California/04/2009 sequences:\n",
      "\tcds:ACP41111 A/California/04/2009 2009/04/01 H1N1 NS2\n",
      "\tcds:ACP44155 A/California/04/2009 2009/04/01 H1N1 NS2\n",
      "The exact protein matches for NS2 include A/California/04/2009 sequences:\n",
      "\tcds:ACP41111 A/California/04/2009 2009/04/01 H1N1 NS2\n",
      "\tcds:ACP44155 A/California/04/2009 2009/04/01 H1N1 NS2\n"
     ]
    }
   ],
   "source": [
    "alignedseqs_by_cdna = dict([(cdna, []) for cdna in cdnas])\n",
    "seqs_d_by_cdna = {}\n",
    "for cdna in cdnas:\n",
    "    seqs = seqs_by_cdna[cdna]\n",
    "    refseq = plasmid_cdnas[cdna]\n",
    "    refprot = str(refseq.translate())\n",
    "    refseq = str(refseq)\n",
    "    idents = []\n",
    "    seqs_d = {refseq:[]}\n",
    "    prots_d = {refseq:[]}\n",
    "    for seq in seqs:\n",
    "        if len(seq.seq) - 3 == len(refseq) and seq.seq[-3 : ] in ['TGA', 'TAA', 'TAG']:\n",
    "            seq.seq = seq.seq[: -3] # trim stop codon\n",
    "        if len(seq.seq) == len(refseq):\n",
    "            ident = len([i for (i, nt_i) in enumerate(refseq) if nt_i == seq[i]]) / float(len(refseq))\n",
    "            idents.append((ident, seq))\n",
    "            alignedseqs_by_cdna[cdna].append(seq)\n",
    "            prot = str(seq.seq.translate())\n",
    "            seqstring = str(seq.seq)\n",
    "            if seqstring in seqs_d:\n",
    "                seqs_d[seqstring].append(seq.description)\n",
    "            else:\n",
    "                seqs_d[seqstring] = [seq.description]\n",
    "            if prot in prots_d:\n",
    "                prots_d[prot].append(seq.description)\n",
    "            else:\n",
    "                prots_d[prot] = [seq.description]\n",
    "    idents.sort()\n",
    "    seqs_d_by_cdna[cdna] = seqs_d\n",
    "    print(\"\\nFor %s, %d of %d sequences aligned gaplessly. The median and minimum nucleotide identities are %.3f and %.3f\" % (cdna, len(alignedseqs_by_cdna[cdna]), len(seqs_by_cdna[cdna]), idents[len(idents) // 2][0], idents[-1][0]))\n",
    "    print(\"The %s nucleotide sequence in the plasmid exactly matches %s known sequences.\" % (cdna, len(seqs_d[refseq])))\n",
    "    print(\"The %s protein sequence in the plasmid exactly matches %s known sequences.\" % (cdna, len(prots_d[refprot])))\n",
    "    if [head for head in seqs_d[refseq] if 'A/California/04/2009' in head]:\n",
    "        print('The exact nucleotide matches for %s include A/California/04/2009 sequences:\\n\\t%s' % (cdna, '\\n\\t'.join([head for head in seqs_d[refseq] if 'A/California/04/2009' in head])))\n",
    "    elif seqs_d[refseq]:\n",
    "        print('The exact nucleotide matches for %s do not include A/California/04/2009 but do include %d other sequences including %s' % (cdna, len(seqs_d[refseq]), seqs_d[refseq][0]))\n",
    "    else:\n",
    "        closestseq = idents[-1][1]\n",
    "        nt_muts = ['%s%d%s' % (nt_i, i + 1, closestseq.seq[i]) for (i, nt_i) in enumerate(refseq) if nt_i != closestseq.seq[i]]\n",
    "        print(\"There are no exact nucleotide matches for %s. The closest sequence is:\\n\\t%s\\n\\tThis sequence has the following nucleotide mutations: %s\" % (cdna, closestseq.description, ', '.join(nt_muts)))\n",
    "    if [head for head in prots_d[refprot] if 'A/California/04/2009' in head]:\n",
    "        print('The exact protein matches for %s include A/California/04/2009 sequences:\\n\\t%s' % (cdna, '\\n\\t'.join([head for head in prots_d[refprot] if 'A/California/04/2009' in head])))\n",
    "    elif prots_d[refprot]:\n",
    "        print('The exact protein matches for %s do not include A/California/04/2009 but do include %d other sequences including %s' % (cdna, len(prots_d[refprot]), prots_d[refprot][0]))\n",
    "    else:\n",
    "        print(\"There are NO exact protein matches for %s\" % cdna)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How closely do we expect the reference sequences to match other sequences?\n",
    "The analysis above finds that some of the reference (plasmid) nucleotide sequences don't exactly match any other nucleotide sequences in the [Influenza Virus Resource]. Is this surprising? Should we expect a \"valid\" pdmH1N1 influenza sequence to exactly match other nucleotide sequences in the database?\n",
    "\n",
    "Below we count how many times each of the sequences in the [Influenza Virus Resource] is found for each cDNA. The results below show that is not unusual for a sequence to have no other exact nucleotide matches in the database (this is true for all cDNAs except *NS2* and *M2*). Therefore, we probably shouldn't be concerned that some of the segments (specifically, *PB1*, *PA*, *NS1*) don't exactly match any other sequence in the [Influenza Virus Resource], since this is true for many sequences for these segments. This result simply indicates that the high genetic diversity of influenza means that most segments will tend to have a lot of genetic variants. We can draw confidence from the fact that the **protein** sequences for all of the reference (plasmid) sequences do match other protein sequences, and so probably don't contain harmful mutations.\n",
    "\n",
    "[Influenza Virus Resource]: http://www.ncbi.nlm.nih.gov/genomes/FLU/FLU.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "For PB2, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t2630 sequences are found no more than 1 times (this is 0.515 of the total)\n",
      "\t3170 sequences are found no more than 2 times (this is 0.620 of the total)\n",
      "\t1941 sequences are found more than 2 times (this is 0.380 of the total)\n",
      "\n",
      "For PB1, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t2572 sequences are found no more than 1 times (this is 0.504 of the total)\n",
      "\t3178 sequences are found no more than 2 times (this is 0.623 of the total)\n",
      "\t1923 sequences are found more than 2 times (this is 0.377 of the total)\n",
      "\n",
      "For PA, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t2620 sequences are found no more than 1 times (this is 0.489 of the total)\n",
      "\t3250 sequences are found no more than 2 times (this is 0.607 of the total)\n",
      "\t2104 sequences are found more than 2 times (this is 0.393 of the total)\n",
      "\n",
      "For HA, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t6166 sequences are found no more than 1 times (this is 0.547 of the total)\n",
      "\t7280 sequences are found no more than 2 times (this is 0.646 of the total)\n",
      "\t3992 sequences are found more than 2 times (this is 0.354 of the total)\n",
      "\n",
      "For NP, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t1715 sequences are found no more than 1 times (this is 0.314 of the total)\n",
      "\t2275 sequences are found no more than 2 times (this is 0.417 of the total)\n",
      "\t3180 sequences are found more than 2 times (this is 0.583 of the total)\n",
      "\n",
      "For NA, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t3787 sequences are found no more than 1 times (this is 0.421 of the total)\n",
      "\t4699 sequences are found no more than 2 times (this is 0.523 of the total)\n",
      "\t4290 sequences are found more than 2 times (this is 0.477 of the total)\n",
      "\n",
      "For M1, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t1082 sequences are found no more than 1 times (this is 0.142 of the total)\n",
      "\t1540 sequences are found no more than 2 times (this is 0.202 of the total)\n",
      "\t6068 sequences are found more than 2 times (this is 0.798 of the total)\n",
      "\n",
      "For M2, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t334 sequences are found no more than 1 times (this is 0.047 of the total)\n",
      "\t458 sequences are found no more than 2 times (this is 0.065 of the total)\n",
      "\t6574 sequences are found more than 2 times (this is 0.935 of the total)\n",
      "\n",
      "For NS1, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t1142 sequences are found no more than 1 times (this is 0.194 of the total)\n",
      "\t1542 sequences are found no more than 2 times (this is 0.262 of the total)\n",
      "\t4353 sequences are found more than 2 times (this is 0.738 of the total)\n",
      "\n",
      "For NS2, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\n",
      "\t337 sequences are found no more than 1 times (this is 0.061 of the total)\n",
      "\t485 sequences are found no more than 2 times (this is 0.088 of the total)\n",
      "\t5052 sequences are found more than 2 times (this is 0.912 of the total)\n"
     ]
    }
   ],
   "source": [
    "for cdna in cdnas:\n",
    "    nmatches = []\n",
    "    for (seq, exactmatches) in seqs_d_by_cdna[cdna].items():\n",
    "        nmatches += [len(exactmatches)] * len(exactmatches)\n",
    "    runningcount = 0\n",
    "    print(\"\\nFor %s, here are the number of sequences with no more than the indicated number of exact nucleotide matches:\" % cdna)\n",
    "    for i in range(1, 3):\n",
    "        runningcount += nmatches.count(i)\n",
    "        print(\"\\t%d sequences are found no more than %d times (this is %.3f of the total)\" % (runningcount, i, runningcount / float(len(nmatches))))\n",
    "    print(\"\\t%d sequences are found more than %d times (this is %.3f of the total)\" % (len(nmatches) - runningcount, i, 1.0 - runningcount / float(len(nmatches))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Design of synonymously barcoded variants\n",
    "We will now design synonymously barcoded variants of all of the genes. We will choose barcodes that do not alter the protein sequence, which change at least two nearby nucleotides (to facilitate Illumina short-read sequencing), and which only make mutations to nucleotides that are quite common in the existing database set of pdmH1N1 sequences. The rationale is that if the synonymous mutations are common in naturally occurring sequences, then they are probably not deleterious.\n",
    "\n",
    "We will attempt to put the mutations somewhat close the 3' end of the gene, in case we ever want to do polyA-based RNAseq. However, we will not put them too close as nucleotides in the noncoding termini are sometimes involved in genome packaging. For the *M* and *NS* segments, we will put the mutations in the *M2* and *NS2* coding regions (after their overlapping regions with *M1* and *NS1*).\n",
    "\n",
    "Below is the design; at the end of the output are the created sequences and the names of the files to which they are written:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "For PB2, found 38 synonymous mutations that occur in at least 0.02 of sequences (this is 102 sequences)\n",
      "We will make the following two codon mutations to PB2: AAG721AAA and CAA728CAG\n",
      "\n",
      "For PB1, found 39 synonymous mutations that occur in at least 0.02 of sequences (this is 102 sequences)\n",
      "We will make the following two codon mutations to PB1: TTC730TTT and AAA737AAG\n",
      "\n",
      "For PA, found 40 synonymous mutations that occur in at least 0.02 of sequences (this is 107 sequences)\n",
      "We will make the following two codon mutations to PA: GAG597GAA and ACC608ACT\n",
      "\n",
      "For HA, found 30 synonymous mutations that occur in at least 0.02 of sequences (this is 225 sequences)\n",
      "We will make the following two codon mutations to HA: CTA470TTA and AAC479AAT\n",
      "\n",
      "For NP, found 28 synonymous mutations that occur in at least 0.02 of sequences (this is 109 sequences)\n",
      "We will make the following two codon mutations to NP: GAA454GAG and TTC458TTT\n",
      "\n",
      "For NA, found 22 synonymous mutations that occur in at least 0.02 of sequences (this is 179 sequences)\n",
      "We will make the following two codon mutations to NA: GTT407GTC and GAT416GAC\n",
      "\n",
      "For M2, found 3 synonymous mutations that occur in at least 0.02 of sequences (this is 140 sequences)\n",
      "We will make the following two codon mutations to M2: GTG68GTA and GAA79GAG\n",
      "\n",
      "For NS2, found 4 synonymous mutations that occur in at least 0.02 of sequences (this is 110 sequences)\n",
      "We will make the following two codon mutations to NS2: AGG77AGA and CGG84CGA\n",
      "\n",
      "Writing the recoded cDNAs to ./sequences/recoded_cDNAs.txt\n",
      "\n",
      "Writing the recoded plasmids to ./sequences/recoded_plasmids.txt\n",
      "\n",
      "\n",
      "Here are the recoded cDNAs in ./sequences/recoded_cDNAs.txt:\n",
      "\n",
      ">PB2 synonymously barcoded with codon mutations AAG721AAA and CAA728CAG\n",
      "ATGGAGAGAATAAAAGAACTGAGAGATCTAATGTCGCAGTCCCGCACTCGCGAGATACTCACTAAGACCACTGTGGACCATATGGCCATAATCAAAAAGTACACATCAGGAAGGCAAGAGAAGAACCCCGCACTCAGAATGAAGTGGATGATGGCAATGAGATACCCAATTACAGCAGACAAGAGAATAATGGACATGATTCCAGAGAGGAATGAACAAGGACAAACCCTCTGGAGCAAAACAAACGATGCTGGATCAGACCGAGTGATGGTATCACCTCTGGCCGTAACATGGTGGAATAGGAATGGCCCAACAACAAGTACAGTTCATTACCCTAAGGTATATAAAACTTATTTCGAAAAGGTCGAAAGGTTGAAACATGGTACCTTCGGCCCTGTCCACTTCAGAAATCAAGTTAAAATAAGGAGGAGAGTTGATACAAACCCTGGCCATGCAGATCTCAGTGCCAAGGAGGCACAGGATGTGATTATGGAAGTTGTTTTCCCAAATGAAGTGGGGGCAAGAATACTGACATCAGAGTCACAGCTGGCAATAACAAAAGAGAAGAAAGAAGAGCTCCAGGATTGTAAAATTGCTCCCTTGATGGTGGCGTACATGCTAGAAAGAGAATTGGTCCGTAAAACAAGGTTTCTCCCAGTAGCCGGCGGAACAGGCAGTGTTTATATTGAAGTGTTGCACTTAACCCAAGGGACGTGCTGGGAGCAGATGTACACTCCAGGAGGAGAAGTGAGAAATGATGATGTTGACCAAAGTTTGATTATCGCTGCTAGAAACATAGTAAGAAGAGCAGCAGTGTCAGCAGACCCATTAGCATCTCTCTTGGAAATGTGCCACAGCACACAGATTGGAGGAGTAAGGATGGTGGACATCCTTAGACAGAATCCAACTGAGGAACAAGCCGTAGACATATGCAAGGCAGCAATAGGGTTGAGGATTAGCTCATCTTTCAGTTTTGGTGGGTTCACTTTCAAAAGGACAAGCGGATCATCAGTCAAGAAAGAAGAAGAAGTGCTAACGGGCAACCTCCAAACACTGAAAATAAGAGTACATGAAGGGTATGAAGAATTCACAATGGTTGGGAGAAGAGCAACAGCTATTCTCAGAAAGGCAACCAGGAGATTGATCCAGTTGATAGTAAGCGGGAGAGACGAGCAGTCAATTGCTGAGGCAATAATTGTGGCCATGGTATTCTCACAGGAGGATTGCATGATCAAGGCAGTTAGGGGCGATCTGAACTTTGTCAATAGGGCAAACCAGCGACTGAACCCCATGCACCAACTCTTGAGGCATTTCCAAAAAGATGCAAAAGTGCTTTTCCAGAACTGGGGAATTGAATCCATCGACAATGTGATGGGAATGATCGGAATACTGCCCGACATGACCCCAAGCACGGAGATGTCGCTGAGAGGGATAAGAGTCAGCAAAATGGGAGTAGATGAATACTCCAGCACGGAGAGAGTGGTAGTGAGTATTGACCGATTTTTAAGGGTTAGAGATCAAAGAGGGAACGTACTATTGTCTCCCGAAGAAGTCAGTGAAACGCAAGGAACTGAGAAGTTGACAATAACTTATTCGTCATCAATGATGTGGGAGATCAATGGCCCTGAGTCAGTGCTAGTCAACACTTATCAATGGATAATCAGGAACTGGGAAATTGTGAAAATTCAATGGTCACAAGATCCCACAATGTTATACAACAAAATGGAATTTGAACCATTTCAGTCTCTTGTCCCTAAGGCAACCAGAAGCCGGTACAGTGGATTCGTAAGGACACTGTTCCAGCAAATGCGGGATGTGCTTGGGACATTTGACACTGTCCAAATAATAAAACTTCTCCCCTTTGCTGCTGCCCCACCAGAACAGAGTAGGATGCAATTTTCCTCATTGACTGTGAATGTGAGAGGATCAGGGTTGAGGATACTGGTAAGAGGCAATTCTCCAGTATTCAATTACAACAAGGCAACCAAACGACTTACAGTTCTTGGAAAGGATGCAGGTGCATTGACTGAAGATCCAGATGAAGGCACATCTGGGGTGGAGTCTGCTGTCCTGAGAGGATTTCTCATTTTGGGCAAAGAAGACAAGAGATATGGCCCAGCATTAAGCATCAATGAACTGAGCAATCTTGCAAAAGGAGAGAAaGCTAATGTGCTAATTGGGCAgGGGGACGTAGTGTTGGTAATGAAACGAAAACGGGACTCTAGCATACTTACTGACAGCCAGACAGCGACCAAAAGAATTCGGATGGCCATCAAT\n",
      "\n",
      ">PB1 synonymously barcoded with codon mutations TTC730TTT and AAA737AAG\n",
      "ATGGATGTCAATCCGACTCTACTTTTCCTAAAAATTCCAGCGCAAAATGCCATAAGCACCACATTCCCTTATACTGGAGATCCTCCATACAGCCATGGAACAGGAACAGGATACACCATGGACACAGTAAACAGAACACACCAATACTCAGAAAAGGGAAAGTGGACGACCAACACAGAGACTGGTGCACCCCAGCTCAACCCGATTGATGGACCACTACCTGAGGATAATGAACCAAGTGGGTATGCACAAACAGACTGTGTTCTAGAGGCTATGGCTTTCCTTGAAGAATCCCACCCAGGAATATTTGAGAATTCATGCCTTGAAACAATGGAAGTTGTTCAACAAACAAGGGTAGATAAACTAACTCAAGGTCGCCAGACTTATGATTGGACATTAAACAGAAATCAACCGGCAGCAACTGCATTGGCCAACACCATAGAAGTCTTTAGATCGAATGGCCTAACAGCAAATGAGTCAGGAAGGCTAATAGATTTCTTAAAGGATGTAATGGAATCAATGAACAAAGAGGAAATAGAGATAACAACCCACTTTCAAAGAAAAAGGAGAGTAAGAGACAACATGACCAAGAAGATGGTCACGCAAAGAACAATAGGGAAGAAAAAACAAAGACTGAATAAGAGAGGCTATCTAATAAGAGCACTGACATTAAATACGATGACCAAAGATGCAGAGAGAGGCAAGTTAAAAAGAAGGGCTATCGCAACACCTGGGATGCAGATTAGAGGTTTCGTATACTTTGTTGAAACTTTAGCTAGGAGCATTTGCGAAAAGCTTGAACAGTCTGGGCTCCCAGTAGGGGGCAATGAAAAGAAGGCCAAACTGGCAAATGTTGTGAGAAAGATGATGACTAATTCACAAGACACAGAGATTTCTTTCACAATCACTGGGGACAACACTAAGTGGAATGAAAATCAAAATCCCCGAATGTTCCTGGCGATGATTACATATATCACCAGAAATCAACCCGAGTGGTTCAGAAACATCCTGAGCATGGCACCCATAATGTTCTCAAACAAAATGGCAAGACTAGGGAAAGGGTACATGTTCGAGAGTAAAAGAATGAAGATTCGAACACAAATACCAGCAGAAATGCTAGCAAGCATTGACCTGAAGTACTTCAATGAATCAACAAAGAAGAAAATTGAGAAAATAAGGCCTCTTCTAATAGATGGCACAGCATCACTGAGTCCTGGGATGATGATGGGCATGTTCAACATGCTAAGTACGGTCTTGGGAGTCTCGATACTGAATCTTGGACAAAAGAAATACACCAAGACAATATACTGGTGGGATGGGCTCCAATCATCCGACGATTTTGCTCTCATAGTGAATGCACCAAACCATGAGGGAATACAAGCAGGAGTGGACAGATTCTACAGGACCTGCAAGTTAGTGGGAATCAACATGAGCAAAAAGAAGTCCTATATAAATAAGACAGGGACATTTGAATTCACAAGCTTTTTTTATCGCTATGGATTTGTGGCTAATTTTAGCATGGAGCTACCCAGCTTTGGAGTGTCTGGAGTAAATGAATCAGCTGACATGAGTATTGGAGTAACAGTGATAAAGAACAACATGATAAACAATGACCTTGGACCTGCAACGGCCCAGATGGCTCTTCAATTGTTCATCAAAGACTACAGATACACATATAGGTGCCATAGGGGAGACACACAAATTCAGACGAGAAGATCATTTGAGTTAAAGAAGCTGTGGGATCAAACCCAATCAAAGGTAGGGCTATTAGTATCAGATGGAGGACCAAACTTATACAATATACGGAATCTTCACATTCCTGAAGTCTGCTTAAAATGGGAGCTAATGGATGATGATTATCGGGGAAGACTTTGTAATCCCCTGAATCCCTTTGTCAGTCATAAAGAGATTGATTCTGTAAACAATGCTGTGGTAATGCCAGCCCATGGTCCAGCCAAAAGCATGGAATATGATGCCGTTGCAACTACACATTCCTGGATTCCCAAGAGGAATCGTTCTATTCTCAACACAAGCCAAAGGGGAATTCTTGAGGATGAACAGATGTACCAGAAGTGCTGCAATCTATTCGAGAAATTTTTCCCTAGCAGTTCATATAGGAGACCGGTTGGAATTTCTAGCATGGTGGAGGCCATGGTGTCTAGGGCCCGGATTGATGCCAGGGTCGACTTtGAGTCTGGACGGATCAAGAAgGAAGAGTTCTCTGAGATCATGAAGATCTGTTCCACCATTGAAGAACTCAGACGGCAAAAA\n",
      "\n",
      ">PA synonymously barcoded with codon mutations GAG597GAA and ACC608ACT\n",
      "ATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATATGGGGAAGATCCGAAAATCGAAACTAACAAGTTTGCTGCAATATGCACACATTTGGAAGTTTGTTTCATGTATTCGGATTTCCATTTCATCGACGAACGGGGTGAATCAATAATTGTAGAATCTGGTGACCCGAATGCACTATTGAAGCACCGATTTGAGATAATTGAAGGAAGAGACCGAATCATGGCCTGGACAGTGGTGAACAGTATATGTAACACAACAGGGGTAGAGAAGCCTAAATTTCTTCCTGATTTGTATGATTACAAAGAGAACCGGTTCATTGAAATTGGAGTAACACGGAGGGAAGTCCACATATATTACCTAGAGAAAGCCAACAAAATAAAATCTGAGAAGACACACATTCACATCTTTTCATTCACTGGAGAGGAGATGGCCACCAAAGCGGACTACACCCTTGACGAAGAGAGCAGGGCAAGAATCAAAACTAGGCTTTTCACTATAAGACAAGAAATGGCCAGTAGGAGTCTATGGGATTCCTTTCGTCAGTCCGAAAGAGGCGAAGAGACAATTGAAGAAAAATTTGAGATTACAGGAACTATGCGCAAGCTTGCCGACCAAAGTCTCCCACCGAACTTCCCCAGCCTTGAAAACTTTAGAGCCTATGTAGATGGATTCGAGCCGAACGGCTGCATTGAGGGCAAGCTTTCCCAAATGTCAAAAGAAGTGAACGCCAAAATTGAACCATTCTTGAGGACGACACCACGCCCCCTCAGATTGCCTGATGGGCCTCTTTGCCATCAGCGGTCAAAGTTCCTGCTGATGGATGCTCTGAAATTAAGTATTGAAGACCCGAGTCACGAGGGGGAGGGAATACCACTATATGATGCAATCAAATGCATGAAGACATTCTTTGGCTGGAAAGAGCCTAACATAGTCAAACCACATGAGAAAGGCATAAATCCTAATTACCTCATGGCTTGGAAGCAGGTGCTAGCAGAGCTACAGGACATTGAAAATGAAGAGAAGATCCCAAGGACAAAGAACATGAAGAGAACAAGCCAATTGAAGTGGGCACTCGGTGAAAATATGGCACCAGAAAAAGTAGACTTTGATGACTGCAAAGATGTTGGAGACCTTAAACAGTATGACAGTGATGAGCCAGAGCCCAGATCTCTAGCAAGCTGGGTCCAAAATGAATTCAATAAGGCATGTGAATTGACTGATTCAAGCTGGATAGAACTTGATGAAATAGGAGAAGATGTTGCCCCGATTGAACATATCGCAAGCATGAGGAGGAACTATTTTACAGCAGAAGTGTCCCACTGCAGGGCTACTGAATACATAATGAAGGGAGTGTACATAAATACGGCCTTGCTCAATGCATCCTGTGCAGCCATGGATGACTTTCAGCTGATCCCAATGATAAGCAAATGTAGGACCAAAGAAGGAAGACGGAAAACAAACCTGTATGGGTTCATTATAAAAGGAAGGTCTCATTTGAGAAATGATACTGATGTGGTGAACTTTGTAAGTATGGAGTTCTCACTCACTGACCCGAGACTGGAGCCACACAAATGGGAAAAATACTGTGTTCTTGAAATAGGAGACATGCTCTTGAGGACTGCGATAGGCCAAGTGTCGAGGCCCATGTTCCTATATGTGAGAACCAATGGAACCTCCAAGATCAAGATGAAATGGGGCATGGAAATGAGGCGCTGCCTTCTTCAGTCTCTTCAGCAGATTGAGAGCATGATTGAaGCCGAGTCTTCTGTCAAAGAGAAAGACATGACtAAGGAATTCTTTGAAAACAAATCGGAAACATGGCCAATCGGAGAGTCACCCAGGGGAGTGGAGGAAGGCTCTATTGGGAAAGTGTGCAGGACCTTACTGGCAAAATCTGTATTCAACAGTCTATATGCGTCTCCACAACTTGAGGGGTTTTCGGCTGAATCTAGAAAATTGCTTCTCATTGTTCAGGCACTTAGGGACAACCTGGAACCTGGAACCTTCGATCTTGGGGGGCTATATGAAGCAATCGAGGAGTGCCTGATTAATGATCCCTGGGTTTTGCTTAATGCATCTTGGTTCAACTCCTTCCTCACACATGCACTGAAG\n",
      "\n",
      ">HA synonymously barcoded with codon mutations CTA470TTA and AAC479AAT\n",
      "ATGAAGGCAATACTAGTAGTTCTGCTATATACATTTGCAACCGCAAATGCAGACACATTATGTATAGGTTATCATGCGAACAATTCAACAGACACTGTAGACACAGTACTAGAAAAGAATGTAACAGTAACACACTCTGTTAACCTTCTAGAAGACAAGCATAACGGGAAACTATGCAAACTAAGAGGGGTAGCCCCATTGCATTTGGGTAAATGTAACATTGCTGGCTGGATCCTGGGAAATCCAGAGTGTGAATCACTCTCCACAGCAAGCTCATGGTCCTACATTGTGGAAACACCTAGTTCAGACAATGGAACGTGTTACCCAGGAGATTTCATCGATTATGAGGAGCTAAGAGAGCAATTGAGCTCAGTGTCATCATTTGAAAGGTTTGAGATATTCCCCAAGACAAGTTCATGGCCCAATCATGACTCGAACAAAGGTGTAACGGCAGCATGTCCTCATGCTGGAGCAAAAAGCTTCTACAAAAATTTAATATGGCTAGTTAAAAAAGGAAATTCATACCCAAAGCTCAGCAAATCCTACATTAATGATAAAGGGAAAGAAGTCCTCGTGCTATGGGGCATTCACCATCCATCTACTAGTGCTGACCAACAAAGTCTCTATCAGAATGCAGATACATATGTTTTTGTGGGGTCATCAAGATACAGCAAGAAGTTCAAGCCGGAAATAGCAATAAGACCCAAAGTGAGGGATCAAGAAGGGAGAATGAACTATTACTGGACACTAGTAGAGCCGGGAGACAAAATAACATTCGAAGCAACTGGAAATCTAGTGGTACCGAGATATGCATTCGCAATGGAAAGAAATGCTGGATCTGGTATTATCATTTCAGATACACCAGTCCACGATTGCAATACAACTTGTCAAACACCCAAGGGTGCTATAAACACCAGCCTCCCATTTCAGAATATACATCCGATCACAATTGGAAAATGTCCAAAATATGTAAAAAGCACAAAATTGAGACTGGCCACAGGATTGAGGAATATCCCGTCTATTCAATCTAGAGGCCTATTTGGGGCCATTGCCGGTTTCATTGAAGGGGGGTGGACAGGGATGGTAGATGGATGGTACGGTTATCACCATCAAAATGAGCAGGGGTCAGGATATGCAGCCGACCTGAAGAGCACACAGAATGCCATTGACGAGATTACTAACAAAGTAAATTCTGTTATTGAAAAGATGAATACACAGTTCACAGCAGTAGGTAAAGAGTTCAACCACCTGGAAAAAAGAATAGAGAATTTAAATAAAAAAGTTGATGATGGTTTCCTGGACATTTGGACTTACAATGCCGAACTGTTGGTTCTATTGGAAAATGAAAGAACTTTGGACTACCACGATTCAAATGTGAAGAACTTATATGAAAAGGTAAGAAGCCAGtTAAAAAACAATGCCAAGGAAATTGGAAAtGGCTGCTTTGAATTTTACCACAAATGCGATAACACGTGCATGGAAAGTGTCAAAAATGGGACTTATGACTACCCAAAATACTCAGAGGAAGCAAAATTAAACAGAGAAGAAATAGATGGGGTAAAGCTGGAATCAACAAGGATTTACCAGATTTTGGCGATCTATTCAACTGTCGCCAGTTCATTGGTACTGGTAGTCTCCCTGGGGGCAATCAGTTTCTGGATGTGCTCTAATGGGTCTCTACAGTGTAGAATATGTATT\n",
      "\n",
      ">NP synonymously barcoded with codon mutations GAA454GAG and TTC458TTT\n",
      "ATGGCGTCTCAAGGCACCAAACGATCATATGAACAAATGGAGACTGGTGGGGAGCGCCAGGATGCCACAGAAATCAGAGCATCTGTCGGAAGAATGATTGGTGGAATCGGGAGATTCTACATCCAAATGTGCACTGAACTCAAACTCAGTGATTATGATGGACGACTAATCCAGAATAGCATAACAATAGAGAGGATGGTGCTTTCTGCTTTTGATGAGAGAAGAAATAAATACCTAGAAGAGCATCCCAGTGCTGGGAAGGACCCTAAGAAAACAGGAGGACCCATATATAGAAGAGTAGACGGAAAGTGGATGAGAGAACTCATCCTTTATGACAAAGAAGAAATAAGGAGAGTTTGGCGCCAAGCAAACAATGGCGAAGATGCAACAGCAGGTCTTACTCATATCATGATTTGGCATTCCAACCTGAATGATGCCACATATCAGAGAACAAGAGCGCTTGTTCGCACCGGAATGGATCCCAGAATGTGCTCTCTAATGCAAGGTTCAACACTTCCCAGAAGGTCTGGTGCCGCAGGTGCTGCGGTGAAAGGAGTTGGAACAATAGCAATGGAGTTAATCAGAATGATCAAACGTGGAATCAATGACCGAAATTTCTGGAGGGGTGAAAATGGACGAAGGACAAGGGTTGCTTATGAAAGAATGTGCAATATCCTCAAAGGAAAATTTCAAACAGCTGCCCAGAGGGCAATGATGGATCAAGTAAGAGAAAGTCGAAACCCAGGAAACGCTGAGATTGAAGACCTCATTTTCCTGGCACGGTCAGCACTCATTCTGAGGGGATCAGTTGCACATAAATCCTGCCTGCCTGCTTGTGTGTATGGGCTTGCAGTAGCAAGTGGGCATGACTTTGAAAGGGAAGGGTACTCACTGGTCGGGATAGACCCATTCAAATTACTCCAAAACAGCCAAGTGGTCAGCCTGATGAGACCAAATGAAAACCCAGCTCACAAGAGTCAATTGGTGTGGATGGCATGCCACTCTGCTGCATTTGAAGATTTAAGAGTATCAAGTTTCATAAGAGGAAAGAAAGTGATTCCAAGAGGAAAGCTTTCCACAAGAGGGGTCCAGATTGCTTCAAATGAGAATGTGGAAACCATGGACTCCAATACCCTGGAACTGAGAAGCAGATACTGGGCCATAAGGACCAGGAGTGGAGGAAATACCAATCAACAAAAGGCATCCGCAGGCCAGATCAGTGTGCAGCCTACATTCTCAGTGCAGCGGAATCTCCCTTTTGAAAGAGCAACCGTTATGGCAGCATTCAGCGGGAACAATGAAGGACGGACATCCGACATGCGAACAGAAGTTATAAGAATGATGGAAAGTGCAAAGCCAGAgGATTTGTCCTTtCAGGGGCGGGGAGTCTTCGAGCTCTCGGACGAAAAGGCAACGAACCCGATCGTGCCTTCCTTTGACATGAGTAATGAAGGGTCTTATTTCTTCGGAGACAATGCAGAGGAGTATGACAGT\n",
      "\n",
      ">NA synonymously barcoded with codon mutations GTT407GTC and GAT416GAC\n",
      "ATGAATCCAAACCAAAAGATAATAACCATTGGTTCGGTCTGTATGACAATTGGAATGGCTAACTTAATATTACAAATTGGAAACATAATCTCAATATGGATTAGCCACTCAATTCAACTTGGGAATCAAAATCAGATTGAAACATGCAATCAAAGCGTCATTACTTATGAAAACAACACTTGGGTAAATCAGACATATGTTAACATCAGCAACACCAACTTTGCTGCTGGACAGTCAGTGGTTTCCGTGAAATTAGCGGGCAATTCCTCTCTCTGCCCTGTTAGTGGATGGGCTATATACAGTAAAGACAACAGTGTAAGAATCGGTTCCAAGGGGGATGTGTTTGTCATAAGGGAACCATTCATATCATGCTCCCCCTTGGAATGCAGAACCTTCTTCTTGACTCAAGGGGCCTTGCTAAATGACAAACATTCCAATGGAACCATTAAAGACAGGAGCCCATATCGAACCCTAATGAGCTGTCCTATTGGTGAAGTTCCCTCTCCATACAACTCAAGATTTGAGTCAGTCGCTTGGTCAGCAAGTGCTTGTCATGATGGCATCAATTGGCTAACAATTGGAATTTCTGGCCCAGACAATGGGGCAGTGGCTGTGTTAAAGTACAACGGCATAATAACAGACACTATCAAGAGTTGGAGAAACAATATATTGAGAACACAAGAGTCTGAATGTGCATGTGTAAATGGTTCTTGCTTTACTGTAATGACCGATGGACCAAGTAATGGACAGGCCTCATACAAGATCTTCAGAATAGAAAAGGGAAAGATAGTCAAATCAGTCGAAATGAATGCCCCTAATTATCACTATGAGGAATGCTCCTGTTATCCTGATTCTAGTGAAATCACATGTGTGTGCAGGGATAACTGGCATGGCTCGAATCGACCGTGGGTGTCTTTCAACCAGAATCTGGAATATCAGATAGGATACATATGCAGTGGGATTTTCGGAGACAATCCACGCCCTAATGATAAGACAGGCAGTTGTGGTCCAGTATCGTCTAATGGAGCAAATGGAGTAAAAGGATTTTCATTCAAATACGGCAATGGTGTTTGGATAGGGAGAACTAAAAGCATTAGTTCAAGAAACGGTTTTGAGATGATTTGGGATCCGAACGGATGGACTGGGACAGACAATAACTTCTCAATAAAGCAAGATATCGTAGGAATAAATGAGTGGTCAGGATATAGCGGGAGTTTTGTcCAGCATCCAGAACTAACAGGGCTGGAcTGTATAAGACCTTGCTTCTGGGTTGAACTAATCAGAGGGCGACCCAAAGAGAACACAATCTGGACTAGCGGGAGCAGCATATCCTTTTGTGGTGTAAACAGTGACACTGTGGGTTGGTCTTGGCCAGACGGTGCTGAGTTGCCATTTACCATTGACAAG\n",
      "\n",
      ">M2 synonymously barcoded with codon mutations GTG68GTA and GAA79GAG\n",
      "ATGAGTCTTCTAACCGAGGTCGAAACGCCTACCAGAAGCGAATGGGAGTGCAGATGCAGCGATTCAAGTGATCCTCTCGTCATTGCAGCAAATATCATTGGGATCTTGCACCTGATATTGTGGATTACTGATCGTCTTTTTTTCAAATGTATTTATCGTCGCTTTAAATACGGTTTGAAAAGAGGGCCTTCTACGGAAGGAGTaCCTGAGTCCATGAGGGAAGAATATCAACAGGAgCAGCAGAGTGCTGTGGATGTTGACGATGGTCATTTTGTCAACATAGAGCTAGAG\n",
      "\n",
      ">NS2 synonymously barcoded with codon mutations AGG77AGA and CGG84CGA\n",
      "ATGGACTCCAACACCATGTCAAGCTTTCAGGACATACTTATGAGGATGTCAAAAATGCAGTTGGGGTCCTCATCGGAGGACTTGAATGGAATGGTAACACGGTTCGAGTCTCTGAAAATATACAGAGATTCGCTTGGAGAAACTGTGATGAGAATGGGAGACCTTCACTACCTCCAGAGCAGAAATGAAAAGTGGCGAGAGCAATTGGGACAGAAATTTGAGGAAATAAGaTGGTTAATTGAAGAAATGCGaCACAGATTGAAAGCGACAGAGAATAGTTTCGAACAAATAACATTTATGCAAGCCTTACAACTACTGCTTGAAGTAGAACAAGAGATAAGAGCTTTCTCGTTTCAGCTTATT\n",
      "\n",
      "Here are the recoded plasmids in ./sequences/recoded_plasmids.txt:\n",
      "\n",
      ">plasmid vRNA insert for cds:AKQ12663 A/Singapore/DMS1/2009 2009/06/12 H1N1 NS2\n",
      "AGCGAAAGCAGGTCAAATATATTCAATATGGAGAGAATAAAAGAACTGAGAGATCTAATGTCGCAGTCCCGCACTCGCGAGATACTCACTAAGACCACTGTGGACCATATGGCCATAATCAAAAAGTACACATCAGGAAGGCAAGAGAAGAACCCCGCACTCAGAATGAAGTGGATGATGGCAATGAGATACCCAATTACAGCAGACAAGAGAATAATGGACATGATTCCAGAGAGGAATGAACAAGGACAAACCCTCTGGAGCAAAACAAACGATGCTGGATCAGACCGAGTGATGGTATCACCTCTGGCCGTAACATGGTGGAATAGGAATGGCCCAACAACAAGTACAGTTCATTACCCTAAGGTATATAAAACTTATTTCGAAAAGGTCGAAAGGTTGAAACATGGTACCTTCGGCCCTGTCCACTTCAGAAATCAAGTTAAAATAAGGAGGAGAGTTGATACAAACCCTGGCCATGCAGATCTCAGTGCCAAGGAGGCACAGGATGTGATTATGGAAGTTGTTTTCCCAAATGAAGTGGGGGCAAGAATACTGACATCAGAGTCACAGCTGGCAATAACAAAAGAGAAGAAAGAAGAGCTCCAGGATTGTAAAATTGCTCCCTTGATGGTGGCGTACATGCTAGAAAGAGAATTGGTCCGTAAAACAAGGTTTCTCCCAGTAGCCGGCGGAACAGGCAGTGTTTATATTGAAGTGTTGCACTTAACCCAAGGGACGTGCTGGGAGCAGATGTACACTCCAGGAGGAGAAGTGAGAAATGATGATGTTGACCAAAGTTTGATTATCGCTGCTAGAAACATAGTAAGAAGAGCAGCAGTGTCAGCAGACCCATTAGCATCTCTCTTGGAAATGTGCCACAGCACACAGATTGGAGGAGTAAGGATGGTGGACATCCTTAGACAGAATCCAACTGAGGAACAAGCCGTAGACATATGCAAGGCAGCAATAGGGTTGAGGATTAGCTCATCTTTCAGTTTTGGTGGGTTCACTTTCAAAAGGACAAGCGGATCATCAGTCAAGAAAGAAGAAGAAGTGCTAACGGGCAACCTCCAAACACTGAAAATAAGAGTACATGAAGGGTATGAAGAATTCACAATGGTTGGGAGAAGAGCAACAGCTATTCTCAGAAAGGCAACCAGGAGATTGATCCAGTTGATAGTAAGCGGGAGAGACGAGCAGTCAATTGCTGAGGCAATAATTGTGGCCATGGTATTCTCACAGGAGGATTGCATGATCAAGGCAGTTAGGGGCGATCTGAACTTTGTCAATAGGGCAAACCAGCGACTGAACCCCATGCACCAACTCTTGAGGCATTTCCAAAAAGATGCAAAAGTGCTTTTCCAGAACTGGGGAATTGAATCCATCGACAATGTGATGGGAATGATCGGAATACTGCCCGACATGACCCCAAGCACGGAGATGTCGCTGAGAGGGATAAGAGTCAGCAAAATGGGAGTAGATGAATACTCCAGCACGGAGAGAGTGGTAGTGAGTATTGACCGATTTTTAAGGGTTAGAGATCAAAGAGGGAACGTACTATTGTCTCCCGAAGAAGTCAGTGAAACGCAAGGAACTGAGAAGTTGACAATAACTTATTCGTCATCAATGATGTGGGAGATCAATGGCCCTGAGTCAGTGCTAGTCAACACTTATCAATGGATAATCAGGAACTGGGAAATTGTGAAAATTCAATGGTCACAAGATCCCACAATGTTATACAACAAAATGGAATTTGAACCATTTCAGTCTCTTGTCCCTAAGGCAACCAGAAGCCGGTACAGTGGATTCGTAAGGACACTGTTCCAGCAAATGCGGGATGTGCTTGGGACATTTGACACTGTCCAAATAATAAAACTTCTCCCCTTTGCTGCTGCCCCACCAGAACAGAGTAGGATGCAATTTTCCTCATTGACTGTGAATGTGAGAGGATCAGGGTTGAGGATACTGGTAAGAGGCAATTCTCCAGTATTCAATTACAACAAGGCAACCAAACGACTTACAGTTCTTGGAAAGGATGCAGGTGCATTGACTGAAGATCCAGATGAAGGCACATCTGGGGTGGAGTCTGCTGTCCTGAGAGGATTTCTCATTTTGGGCAAAGAAGACAAGAGATATGGCCCAGCATTAAGCATCAATGAACTGAGCAATCTTGCAAAAGGAGAGAAaGCTAATGTGCTAATTGGGCAgGGGGACGTAGTGTTGGTAATGAAACGAAAACGGGACTCTAGCATACTTACTGACAGCCAGACAGCGACCAAAAGAATTCGGATGGCCATCAATTAGTGTCGAATTGTTTAAAAACGACCTTGTTTCTACT\n",
      "\n",
      ">plasmid vRNA insert for cds:AKQ12663 A/Singapore/DMS1/2009 2009/06/12 H1N1 NS2\n",
      "AGCGAAAGCAGGCAAACCATTTGAATGGATGTCAATCCGACTCTACTTTTCCTAAAAATTCCAGCGCAAAATGCCATAAGCACCACATTCCCTTATACTGGAGATCCTCCATACAGCCATGGAACAGGAACAGGATACACCATGGACACAGTAAACAGAACACACCAATACTCAGAAAAGGGAAAGTGGACGACCAACACAGAGACTGGTGCACCCCAGCTCAACCCGATTGATGGACCACTACCTGAGGATAATGAACCAAGTGGGTATGCACAAACAGACTGTGTTCTAGAGGCTATGGCTTTCCTTGAAGAATCCCACCCAGGAATATTTGAGAATTCATGCCTTGAAACAATGGAAGTTGTTCAACAAACAAGGGTAGATAAACTAACTCAAGGTCGCCAGACTTATGATTGGACATTAAACAGAAATCAACCGGCAGCAACTGCATTGGCCAACACCATAGAAGTCTTTAGATCGAATGGCCTAACAGCAAATGAGTCAGGAAGGCTAATAGATTTCTTAAAGGATGTAATGGAATCAATGAACAAAGAGGAAATAGAGATAACAACCCACTTTCAAAGAAAAAGGAGAGTAAGAGACAACATGACCAAGAAGATGGTCACGCAAAGAACAATAGGGAAGAAAAAACAAAGACTGAATAAGAGAGGCTATCTAATAAGAGCACTGACATTAAATACGATGACCAAAGATGCAGAGAGAGGCAAGTTAAAAAGAAGGGCTATCGCAACACCTGGGATGCAGATTAGAGGTTTCGTATACTTTGTTGAAACTTTAGCTAGGAGCATTTGCGAAAAGCTTGAACAGTCTGGGCTCCCAGTAGGGGGCAATGAAAAGAAGGCCAAACTGGCAAATGTTGTGAGAAAGATGATGACTAATTCACAAGACACAGAGATTTCTTTCACAATCACTGGGGACAACACTAAGTGGAATGAAAATCAAAATCCCCGAATGTTCCTGGCGATGATTACATATATCACCAGAAATCAACCCGAGTGGTTCAGAAACATCCTGAGCATGGCACCCATAATGTTCTCAAACAAAATGGCAAGACTAGGGAAAGGGTACATGTTCGAGAGTAAAAGAATGAAGATTCGAACACAAATACCAGCAGAAATGCTAGCAAGCATTGACCTGAAGTACTTCAATGAATCAACAAAGAAGAAAATTGAGAAAATAAGGCCTCTTCTAATAGATGGCACAGCATCACTGAGTCCTGGGATGATGATGGGCATGTTCAACATGCTAAGTACGGTCTTGGGAGTCTCGATACTGAATCTTGGACAAAAGAAATACACCAAGACAATATACTGGTGGGATGGGCTCCAATCATCCGACGATTTTGCTCTCATAGTGAATGCACCAAACCATGAGGGAATACAAGCAGGAGTGGACAGATTCTACAGGACCTGCAAGTTAGTGGGAATCAACATGAGCAAAAAGAAGTCCTATATAAATAAGACAGGGACATTTGAATTCACAAGCTTTTTTTATCGCTATGGATTTGTGGCTAATTTTAGCATGGAGCTACCCAGCTTTGGAGTGTCTGGAGTAAATGAATCAGCTGACATGAGTATTGGAGTAACAGTGATAAAGAACAACATGATAAACAATGACCTTGGACCTGCAACGGCCCAGATGGCTCTTCAATTGTTCATCAAAGACTACAGATACACATATAGGTGCCATAGGGGAGACACACAAATTCAGACGAGAAGATCATTTGAGTTAAAGAAGCTGTGGGATCAAACCCAATCAAAGGTAGGGCTATTAGTATCAGATGGAGGACCAAACTTATACAATATACGGAATCTTCACATTCCTGAAGTCTGCTTAAAATGGGAGCTAATGGATGATGATTATCGGGGAAGACTTTGTAATCCCCTGAATCCCTTTGTCAGTCATAAAGAGATTGATTCTGTAAACAATGCTGTGGTAATGCCAGCCCATGGTCCAGCCAAAAGCATGGAATATGATGCCGTTGCAACTACACATTCCTGGATTCCCAAGAGGAATCGTTCTATTCTCAACACAAGCCAAAGGGGAATTCTTGAGGATGAACAGATGTACCAGAAGTGCTGCAATCTATTCGAGAAATTTTTCCCTAGCAGTTCATATAGGAGACCGGTTGGAATTTCTAGCATGGTGGAGGCCATGGTGTCTAGGGCCCGGATTGATGCCAGGGTCGACTTtGAGTCTGGACGGATCAAGAAgGAAGAGTTCTCTGAGATCATGAAGATCTGTTCCACCATTGAAGAACTCAGACGGCAAAAATAATGAATTTAACTTGTCCTTCATGAAAAAATGCCTTGTTTCTACT\n",
      "\n",
      ">plasmid vRNA insert for cds:AKQ12663 A/Singapore/DMS1/2009 2009/06/12 H1N1 NS2\n",
      "AGCGAAAGCAGGTACTGATCCAAAATGGAAGACTTTGTGCGACAATGCTTCAATCCAATGATCGTCGAGCTTGCGGAAAAGGCAATGAAAGAATATGGGGAAGATCCGAAAATCGAAACTAACAAGTTTGCTGCAATATGCACACATTTGGAAGTTTGTTTCATGTATTCGGATTTCCATTTCATCGACGAACGGGGTGAATCAATAATTGTAGAATCTGGTGACCCGAATGCACTATTGAAGCACCGATTTGAGATAATTGAAGGAAGAGACCGAATCATGGCCTGGACAGTGGTGAACAGTATATGTAACACAACAGGGGTAGAGAAGCCTAAATTTCTTCCTGATTTGTATGATTACAAAGAGAACCGGTTCATTGAAATTGGAGTAACACGGAGGGAAGTCCACATATATTACCTAGAGAAAGCCAACAAAATAAAATCTGAGAAGACACACATTCACATCTTTTCATTCACTGGAGAGGAGATGGCCACCAAAGCGGACTACACCCTTGACGAAGAGAGCAGGGCAAGAATCAAAACTAGGCTTTTCACTATAAGACAAGAAATGGCCAGTAGGAGTCTATGGGATTCCTTTCGTCAGTCCGAAAGAGGCGAAGAGACAATTGAAGAAAAATTTGAGATTACAGGAACTATGCGCAAGCTTGCCGACCAAAGTCTCCCACCGAACTTCCCCAGCCTTGAAAACTTTAGAGCCTATGTAGATGGATTCGAGCCGAACGGCTGCATTGAGGGCAAGCTTTCCCAAATGTCAAAAGAAGTGAACGCCAAAATTGAACCATTCTTGAGGACGACACCACGCCCCCTCAGATTGCCTGATGGGCCTCTTTGCCATCAGCGGTCAAAGTTCCTGCTGATGGATGCTCTGAAATTAAGTATTGAAGACCCGAGTCACGAGGGGGAGGGAATACCACTATATGATGCAATCAAATGCATGAAGACATTCTTTGGCTGGAAAGAGCCTAACATAGTCAAACCACATGAGAAAGGCATAAATCCTAATTACCTCATGGCTTGGAAGCAGGTGCTAGCAGAGCTACAGGACATTGAAAATGAAGAGAAGATCCCAAGGACAAAGAACATGAAGAGAACAAGCCAATTGAAGTGGGCACTCGGTGAAAATATGGCACCAGAAAAAGTAGACTTTGATGACTGCAAAGATGTTGGAGACCTTAAACAGTATGACAGTGATGAGCCAGAGCCCAGATCTCTAGCAAGCTGGGTCCAAAATGAATTCAATAAGGCATGTGAATTGACTGATTCAAGCTGGATAGAACTTGATGAAATAGGAGAAGATGTTGCCCCGATTGAACATATCGCAAGCATGAGGAGGAACTATTTTACAGCAGAAGTGTCCCACTGCAGGGCTACTGAATACATAATGAAGGGAGTGTACATAAATACGGCCTTGCTCAATGCATCCTGTGCAGCCATGGATGACTTTCAGCTGATCCCAATGATAAGCAAATGTAGGACCAAAGAAGGAAGACGGAAAACAAACCTGTATGGGTTCATTATAAAAGGAAGGTCTCATTTGAGAAATGATACTGATGTGGTGAACTTTGTAAGTATGGAGTTCTCACTCACTGACCCGAGACTGGAGCCACACAAATGGGAAAAATACTGTGTTCTTGAAATAGGAGACATGCTCTTGAGGACTGCGATAGGCCAAGTGTCGAGGCCCATGTTCCTATATGTGAGAACCAATGGAACCTCCAAGATCAAGATGAAATGGGGCATGGAAATGAGGCGCTGCCTTCTTCAGTCTCTTCAGCAGATTGAGAGCATGATTGAaGCCGAGTCTTCTGTCAAAGAGAAAGACATGACtAAGGAATTCTTTGAAAACAAATCGGAAACATGGCCAATCGGAGAGTCACCCAGGGGAGTGGAGGAAGGCTCTATTGGGAAAGTGTGCAGGACCTTACTGGCAAAATCTGTATTCAACAGTCTATATGCGTCTCCACAACTTGAGGGGTTTTCGGCTGAATCTAGAAAATTGCTTCTCATTGTTCAGGCACTTAGGGACAACCTGGAACCTGGAACCTTCGATCTTGGGGGGCTATATGAAGCAATCGAGGAGTGCCTGATTAATGATCCCTGGGTTTTGCTTAATGCATCTTGGTTCAACTCCTTCCTCACACATGCACTGAAGTAGTTGTGGCAATGCTACTATTTGCTATCCATACTGTCCAAAAAAGTACCTTGTTTCTACT\n",
      "\n",
      ">plasmid vRNA insert for cds:AKQ12663 A/Singapore/DMS1/2009 2009/06/12 H1N1 NS2\n",
      "AGCAAAAGCAGGGGAAAACAAAAGCAACAAAAATGAAGGCAATACTAGTAGTTCTGCTATATACATTTGCAACCGCAAATGCAGACACATTATGTATAGGTTATCATGCGAACAATTCAACAGACACTGTAGACACAGTACTAGAAAAGAATGTAACAGTAACACACTCTGTTAACCTTCTAGAAGACAAGCATAACGGGAAACTATGCAAACTAAGAGGGGTAGCCCCATTGCATTTGGGTAAATGTAACATTGCTGGCTGGATCCTGGGAAATCCAGAGTGTGAATCACTCTCCACAGCAAGCTCATGGTCCTACATTGTGGAAACACCTAGTTCAGACAATGGAACGTGTTACCCAGGAGATTTCATCGATTATGAGGAGCTAAGAGAGCAATTGAGCTCAGTGTCATCATTTGAAAGGTTTGAGATATTCCCCAAGACAAGTTCATGGCCCAATCATGACTCGAACAAAGGTGTAACGGCAGCATGTCCTCATGCTGGAGCAAAAAGCTTCTACAAAAATTTAATATGGCTAGTTAAAAAAGGAAATTCATACCCAAAGCTCAGCAAATCCTACATTAATGATAAAGGGAAAGAAGTCCTCGTGCTATGGGGCATTCACCATCCATCTACTAGTGCTGACCAACAAAGTCTCTATCAGAATGCAGATACATATGTTTTTGTGGGGTCATCAAGATACAGCAAGAAGTTCAAGCCGGAAATAGCAATAAGACCCAAAGTGAGGGATCAAGAAGGGAGAATGAACTATTACTGGACACTAGTAGAGCCGGGAGACAAAATAACATTCGAAGCAACTGGAAATCTAGTGGTACCGAGATATGCATTCGCAATGGAAAGAAATGCTGGATCTGGTATTATCATTTCAGATACACCAGTCCACGATTGCAATACAACTTGTCAAACACCCAAGGGTGCTATAAACACCAGCCTCCCATTTCAGAATATACATCCGATCACAATTGGAAAATGTCCAAAATATGTAAAAAGCACAAAATTGAGACTGGCCACAGGATTGAGGAATATCCCGTCTATTCAATCTAGAGGCCTATTTGGGGCCATTGCCGGTTTCATTGAAGGGGGGTGGACAGGGATGGTAGATGGATGGTACGGTTATCACCATCAAAATGAGCAGGGGTCAGGATATGCAGCCGACCTGAAGAGCACACAGAATGCCATTGACGAGATTACTAACAAAGTAAATTCTGTTATTGAAAAGATGAATACACAGTTCACAGCAGTAGGTAAAGAGTTCAACCACCTGGAAAAAAGAATAGAGAATTTAAATAAAAAAGTTGATGATGGTTTCCTGGACATTTGGACTTACAATGCCGAACTGTTGGTTCTATTGGAAAATGAAAGAACTTTGGACTACCACGATTCAAATGTGAAGAACTTATATGAAAAGGTAAGAAGCCAGtTAAAAAACAATGCCAAGGAAATTGGAAAtGGCTGCTTTGAATTTTACCACAAATGCGATAACACGTGCATGGAAAGTGTCAAAAATGGGACTTATGACTACCCAAAATACTCAGAGGAAGCAAAATTAAACAGAGAAGAAATAGATGGGGTAAAGCTGGAATCAACAAGGATTTACCAGATTTTGGCGATCTATTCAACTGTCGCCAGTTCATTGGTACTGGTAGTCTCCCTGGGGGCAATCAGTTTCTGGATGTGCTCTAATGGGTCTCTACAGTGTAGAATATGTATTTAACATTAGGATTTCAGAAGCATGAGAAAAACACCCTTGTTCTACT\n",
      "\n",
      ">plasmid vRNA insert for cds:AKQ12663 A/Singapore/DMS1/2009 2009/06/12 H1N1 NS2\n",
      "AGCAAAAGCAGGGTAGATAATCACTCAATGAGTGACATCGAAGCCATGGCGTCTCAAGGCACCAAACGATCATATGAACAAATGGAGACTGGTGGGGAGCGCCAGGATGCCACAGAAATCAGAGCATCTGTCGGAAGAATGATTGGTGGAATCGGGAGATTCTACATCCAAATGTGCACTGAACTCAAACTCAGTGATTATGATGGACGACTAATCCAGAATAGCATAACAATAGAGAGGATGGTGCTTTCTGCTTTTGATGAGAGAAGAAATAAATACCTAGAAGAGCATCCCAGTGCTGGGAAGGACCCTAAGAAAACAGGAGGACCCATATATAGAAGAGTAGACGGAAAGTGGATGAGAGAACTCATCCTTTATGACAAAGAAGAAATAAGGAGAGTTTGGCGCCAAGCAAACAATGGCGAAGATGCAACAGCAGGTCTTACTCATATCATGATTTGGCATTCCAACCTGAATGATGCCACATATCAGAGAACAAGAGCGCTTGTTCGCACCGGAATGGATCCCAGAATGTGCTCTCTAATGCAAGGTTCAACACTTCCCAGAAGGTCTGGTGCCGCAGGTGCTGCGGTGAAAGGAGTTGGAACAATAGCAATGGAGTTAATCAGAATGATCAAACGTGGAATCAATGACCGAAATTTCTGGAGGGGTGAAAATGGACGAAGGACAAGGGTTGCTTATGAAAGAATGTGCAATATCCTCAAAGGAAAATTTCAAACAGCTGCCCAGAGGGCAATGATGGATCAAGTAAGAGAAAGTCGAAACCCAGGAAACGCTGAGATTGAAGACCTCATTTTCCTGGCACGGTCAGCACTCATTCTGAGGGGATCAGTTGCACATAAATCCTGCCTGCCTGCTTGTGTGTATGGGCTTGCAGTAGCAAGTGGGCATGACTTTGAAAGGGAAGGGTACTCACTGGTCGGGATAGACCCATTCAAATTACTCCAAAACAGCCAAGTGGTCAGCCTGATGAGACCAAATGAAAACCCAGCTCACAAGAGTCAATTGGTGTGGATGGCATGCCACTCTGCTGCATTTGAAGATTTAAGAGTATCAAGTTTCATAAGAGGAAAGAAAGTGATTCCAAGAGGAAAGCTTTCCACAAGAGGGGTCCAGATTGCTTCAAATGAGAATGTGGAAACCATGGACTCCAATACCCTGGAACTGAGAAGCAGATACTGGGCCATAAGGACCAGGAGTGGAGGAAATACCAATCAACAAAAGGCATCCGCAGGCCAGATCAGTGTGCAGCCTACATTCTCAGTGCAGCGGAATCTCCCTTTTGAAAGAGCAACCGTTATGGCAGCATTCAGCGGGAACAATGAAGGACGGACATCCGACATGCGAACAGAAGTTATAAGAATGATGGAAAGTGCAAAGCCAGAgGATTTGTCCTTtCAGGGGCGGGGAGTCTTCGAGCTCTCGGACGAAAAGGCAACGAACCCGATCGTGCCTTCCTTTGACATGAGTAATGAAGGGTCTTATTTCTTCGGAGACAATGCAGAGGAGTATGACAGTTGAGGAAAAATACCCTTGTTTCTACT\n",
      "\n",
      ">plasmid vRNA insert for cds:AKQ12663 A/Singapore/DMS1/2009 2009/06/12 H1N1 NS2\n",
      "AGCAAAAGCAGGAGTTTAAAATGAATCCAAACCAAAAGATAATAACCATTGGTTCGGTCTGTATGACAATTGGAATGGCTAACTTAATATTACAAATTGGAAACATAATCTCAATATGGATTAGCCACTCAATTCAACTTGGGAATCAAAATCAGATTGAAACATGCAATCAAAGCGTCATTACTTATGAAAACAACACTTGGGTAAATCAGACATATGTTAACATCAGCAACACCAACTTTGCTGCTGGACAGTCAGTGGTTTCCGTGAAATTAGCGGGCAATTCCTCTCTCTGCCCTGTTAGTGGATGGGCTATATACAGTAAAGACAACAGTGTAAGAATCGGTTCCAAGGGGGATGTGTTTGTCATAAGGGAACCATTCATATCATGCTCCCCCTTGGAATGCAGAACCTTCTTCTTGACTCAAGGGGCCTTGCTAAATGACAAACATTCCAATGGAACCATTAAAGACAGGAGCCCATATCGAACCCTAATGAGCTGTCCTATTGGTGAAGTTCCCTCTCCATACAACTCAAGATTTGAGTCAGTCGCTTGGTCAGCAAGTGCTTGTCATGATGGCATCAATTGGCTAACAATTGGAATTTCTGGCCCAGACAATGGGGCAGTGGCTGTGTTAAAGTACAACGGCATAATAACAGACACTATCAAGAGTTGGAGAAACAATATATTGAGAACACAAGAGTCTGAATGTGCATGTGTAAATGGTTCTTGCTTTACTGTAATGACCGATGGACCAAGTAATGGACAGGCCTCATACAAGATCTTCAGAATAGAAAAGGGAAAGATAGTCAAATCAGTCGAAATGAATGCCCCTAATTATCACTATGAGGAATGCTCCTGTTATCCTGATTCTAGTGAAATCACATGTGTGTGCAGGGATAACTGGCATGGCTCGAATCGACCGTGGGTGTCTTTCAACCAGAATCTGGAATATCAGATAGGATACATATGCAGTGGGATTTTCGGAGACAATCCACGCCCTAATGATAAGACAGGCAGTTGTGGTCCAGTATCGTCTAATGGAGCAAATGGAGTAAAAGGATTTTCATTCAAATACGGCAATGGTGTTTGGATAGGGAGAACTAAAAGCATTAGTTCAAGAAACGGTTTTGAGATGATTTGGGATCCGAACGGATGGACTGGGACAGACAATAACTTCTCAATAAAGCAAGATATCGTAGGAATAAATGAGTGGTCAGGATATAGCGGGAGTTTTGTcCAGCATCCAGAACTAACAGGGCTGGAcTGTATAAGACCTTGCTTCTGGGTTGAACTAATCAGAGGGCGACCCAAAGAGAACACAATCTGGACTAGCGGGAGCAGCATATCCTTTTGTGGTGTAAACAGTGACACTGTGGGTTGGTCTTGGCCAGACGGTGCTGAGTTGCCATTTACCATTGACAAGTAATTTGTTCAAAAAACTCCTTGTTTCTACT\n",
      "\n",
      ">plasmid vRNA insert for cds:AKQ12663 A/Singapore/DMS1/2009 2009/06/12 H1N1 NS2\n",
      "AGCAAAAGCAGGTAGATATTTAAAGATGAGTCTTCTAACCGAGGTCGAAACGTACGTTCTTTCTATCATCCCGTCAGGCCCCCTCAAAGCCGAGATCGCGCAGAGACTGGAAAGTGTCTTTGCAGGAAAGAACACAGATCTTGAGGCTCTCATGGAATGGCTAAAGACAAGACCAATCTTGTCACCTCTGACTAAGGGAATTTTAGGATTTGTGTTCACGCTCACCGTGCCCAGTGAGCGAGGACTGCAGCGTAGACGCTTTGTCCAAAATGCCCTAAATGGGAATGGGGACCCGAACAACATGGATAGAGCAGTTAAACTATACAAGAAGCTCAAAAGAGAAATAACGTTCCATGGGGCCAAGGAGGTGTCACTAAGCTATTCAACTGGTGCACTTGCCAGTTGCATGGGCCTCATATACAACAGGATGGGAACAGTGACCACAGAAGCTGCTTTTGGTCTAGTGTGTGCCACTTGTGAACAGATTGCTGATTCACAGCATCGGTCTCACAGACAGATGGCTACTACCACCAATCCACTAATCAGGCATGAAAACAGAATGGTGCTGGCTAGCACTACGGCAAAGGCTATGGAACAGATGGCTGGATCGAGTGAACAGGCAGCGGAGGCCATGGAGGTTGCTAATCAGACTAGGCAGATGGTACATGCAATGAGAACTATTGGGACTCATCCTAGCTCCAGTGCTGGTCTGAAAGATGACCTTCTTGAAAATTTGCAGGCCTACCAGAAGCGAATGGGAGTGCAGATGCAGCGATTCAAGTGATCCTCTCGTCATTGCAGCAAATATCATTGGGATCTTGCACCTGATATTGTGGATTACTGATCGTCTTTTTTTCAAATGTATTTATCGTCGCTTTAAATACGGTTTGAAAAGAGGGCCTTCTACGGAAGGAGTaCCTGAGTCCATGAGGGAAGAATATCAACAGGAgCAGCAGAGTGCTGTGGATGTTGACGATGGTCATTTTGTCAACATAGAGCTAGAGTAAAAAACTACCTTGTTCTACT\n",
      "\n",
      ">plasmid vRNA insert for cds:AKQ12663 A/Singapore/DMS1/2009 2009/06/12 H1N1 NS2\n",
      "AGCAAAAGCAGGGTGACAAAAACATAATGGACTCCAACACCATGTCAAGCTTTCAGGTAGACTGTTTCCTTTGGCATATCCGCAAGCGATTTGCAGACAATGGATTGGGTGATGCCCCATTCCTTGATCGGCTCCGCCGAGATCAAAAGTCCTTAAAAGGAAGAGGCAACACCCTTGGCCTCGATATCGAAACAGCCACTCTTGTTGGGAAACAAATCGTGGAATGGATCTTGAAAGAGGAATCCAGCGAGACACTTAGAATGACAATTGCATCTGTACCTACTTCGCGTTACCTTTCTGACATGACCCTCGAGGAAATGTCACGAGACTGGTTCATGCTCATGCCTAGGCAAAAGATAATAGGCCCTCTTTGCGTGCGATTGGACCAGGCGATCATGGAAAAGAACATAGTACTGAAAGCGAACTTCAGTGTAATCTTTAACCGATTAGAGACCTTGATACTACTAAGGGCTTTCACTGAGGAGGGAGCAATAGTTGGAGAAATTTCACCATTACCTTCTCTTCCAGGACATACTTATGAGGATGTCAAAAATGCAGTTGGGGTCCTCATCGGAGGACTTGAATGGAATGGTAACACGGTTCGAGTCTCTGAAAATATACAGAGATTCGCTTGGAGAAACTGTGATGAGAATGGGAGACCTTCACTACCTCCAGAGCAGAAATGAAAAGTGGCGAGAGCAATTGGGACAGAAATTTGAGGAAATAAGaTGGTTAATTGAAGAAATGCGaCACAGATTGAAAGCGACAGAGAATAGTTTCGAACAAATAACATTTATGCAAGCCTTACAACTACTGCTTGAAGTAGAACAAGAGATAAGAGCTTTCTCGTTTCAGCTTATTTAATGATAAAAAACACCCTTGTTTCTACT"
     ]
    }
   ],
   "source": [
    "# we will take the first synonymous mutation that meets the following criteria:\n",
    "maxdistance = 33 # look mutations to be within this many nucleotides of each other\n",
    "minfreq = 0.02 # require mutations to be present in at least this fraction of all sequences\n",
    "minstopdistance = 50 # require mutations to be at least this many nucleotides upstream of stop codon\n",
    "\n",
    "# now find such mutations...\n",
    "import math\n",
    "recoded_cdnas = []\n",
    "recoded_plasmids = []\n",
    "for cdna in cdnas:\n",
    "    if cdna in ['NS1', 'M1']:\n",
    "        continue # we barcode in the NS2 / M2 regions instead\n",
    "    refseq = plasmid_cdnas[cdna]\n",
    "    alignedseqs = alignedseqs_by_cdna[cdna]\n",
    "    mincounts = len(alignedseqs) * minfreq # only consider synonymous mutations that occur at least this many times\n",
    "    syn_muts = [] # for each codon site with synonymous mutation that occurs >= mincounts time, is (site, most common syn mutation)\n",
    "    for icodon in range(len(refseq) // 3):\n",
    "        wtcodon = refseq[icodon * 3 : icodon * 3 + 3]\n",
    "        codon_muts = {}\n",
    "        for seq in alignedseqs:\n",
    "            mutcodon = seq.seq[icodon * 3 : icodon * 3 + 3]\n",
    "            if str(mutcodon) != str(wtcodon) and str(wtcodon.translate()) == str(mutcodon.translate()):\n",
    "                if mutcodon in codon_muts:\n",
    "                    codon_muts[mutcodon] += 1\n",
    "                else:\n",
    "                    codon_muts[mutcodon] = 1\n",
    "        codon_muts = [(n, mutcodon) for (mutcodon, n) in codon_muts.items()]\n",
    "        codon_muts.sort()\n",
    "        if codon_muts and codon_muts[-1][0] >= mincounts:\n",
    "            syn_muts.append((icodon + 1, wtcodon, codon_muts[-1][1]))\n",
    "    print(\"\\nFor %s, found %d synonymous mutations that occur in at least %.2f of sequences (this is %d sequences)\" % (cdna, len(syn_muts), minfreq, mincounts))\n",
    "    syn_muts.reverse()\n",
    "    lastallowedcodon = int(math.ceil(len(refseq) - minstopdistance) / 3.)\n",
    "    if cdna == 'NS2':\n",
    "        # make sure site does not overlap with NS1 \n",
    "        firstallowedcodon = (len(plasmid_cdnas['NS1']) - splicesites['NS2'][1] + splicesites['NS2'][0]) // 3 + 1\n",
    "    elif cdna == 'M2':\n",
    "        # make sure site does not overlap with M1 \n",
    "        firstallowedcodon = (len(plasmid_cdnas['M1']) - splicesites['M2'][1] + splicesites['M2'][0]) // 3 + 1\n",
    "    else:\n",
    "        firstallowedcodon = 1\n",
    "    for (i, (icodon, wtcodon, mutcodon)) in enumerate(syn_muts[ : -1]):\n",
    "        if icodon > lastallowedcodon or icodon < firstallowedcodon:\n",
    "            continue # too close to 3' end, might be involved in genome packaging\n",
    "        (nexticodon, nextwtcodon, nextmutcodon) = syn_muts[i + 1]\n",
    "        if (icodon - nexticodon) * 3 <= maxdistance: # sufficiently close\n",
    "            break\n",
    "    else:\n",
    "        raise RuntimeError(\"Failed to find synonymous mutations for %s\" % cdna)\n",
    "    (mut1, mut2) = ('%s%d%s' % (nextwtcodon, nexticodon, nextmutcodon), '%s%d%s' % (wtcodon, icodon, mutcodon))\n",
    "    print(\"We will make the following two codon mutations to %s: %s and %s\" % (cdna, mut1, mut2))\n",
    "    recoded = list(str(refseq))\n",
    "    for (i, wt, mut) in [(nexticodon, nextwtcodon, nextmutcodon), (icodon, wtcodon, mutcodon)]:\n",
    "        for (j, nt) in enumerate(wt):\n",
    "            assert recoded[(i - 1) * 3 + j] == nt\n",
    "            if mut[j] != nt:\n",
    "                recoded[(i - 1) * 3 + j] = mut[j].lower()\n",
    "    recoded = ''.join(recoded)\n",
    "    assert len(refseq) == len(recoded)\n",
    "    assert 2 == len([i for i in range(len(recoded)) if recoded[i] != str(refseq)[i]])\n",
    "    newhead = '%s synonymously barcoded with codon mutations %s and %s' % (cdna, mut1, mut2)\n",
    "    recoded_cdnas.append((newhead, recoded))\n",
    "    if cdna in ['NS2', 'M2']:\n",
    "        plasmid = [str(seq.seq) for seq in plasmids if cdna[ : -1] in seq.description]\n",
    "        assert len(plasmid) == 1\n",
    "        plasmid = plasmid[0]\n",
    "        trim5 = splicesites[cdna][0] + 1\n",
    "        i = plasmid.index(str(refseq)[trim5 : ])\n",
    "        recodedplasmid = '%s%s%s' % (plasmid[ : i], recoded[trim5 : ], plasmid[i + len(recoded[trim5 : ]) : ])\n",
    "        assert len(recodedplasmid) == len(plasmid), \"%d vs %d\" % (len(recodedplasmid), len(plasmid))\n",
    "        assert 2 == len([j for j in range(len(plasmid)) if plasmid[j] != recodedplasmid[j]])\n",
    "    else:\n",
    "        plasmid = [str(seq.seq) for seq in plasmids if cdna in seq.description]\n",
    "        assert len(plasmid) == 1\n",
    "        plasmid = plasmid[0]\n",
    "        i = plasmid.index(str(refseq))\n",
    "        recodedplasmid = \"%s%s%s\" % (plasmid[: i], recoded, plasmid[i + len(recoded) : ])\n",
    "        assert len(recodedplasmid) == len(plasmid)\n",
    "        assert 2 == len([j for j in range(len(plasmid)) if plasmid[j] != recodedplasmid[j]])\n",
    "    recoded_plasmids.append(('plasmid vRNA insert for %s' % head, recodedplasmid))\n",
    "\n",
    "# write to files\n",
    "recoded_cdnas_file = '%s/recoded_cDNAs.txt' % seqdir\n",
    "recoded_plasmids_file = '%s/recoded_plasmids.txt' % seqdir\n",
    "for (fname, seqtype, seqlist) in [(recoded_cdnas_file, 'cDNAs', recoded_cdnas), (recoded_plasmids_file, 'plasmids', recoded_plasmids)]:\n",
    "    print(\"\\nWriting the recoded %s to %s\" % (seqtype, fname))\n",
    "    with open(fname, 'w') as f:\n",
    "        f.write('\\n\\n'.join(['>%s\\n%s' % tup for tup in seqlist]))\n",
    "        \n",
    "print('\\n\\nHere are the recoded cDNAs in %s:\\n' % recoded_cdnas_file)\n",
    "!cat $recoded_cdnas_file\n",
    "\n",
    "print('\\n\\nHere are the recoded plasmids in %s:\\n' % recoded_plasmids_file)\n",
    "!cat $recoded_plasmids_file\n",
    "            "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
